internal/telemetry/appinsights.go

```diff
@@ -0,0 +1,67 @@
+package telemetry
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/microsoft/ApplicationInsights-Go/appinsights"
+)
+
+var telemetryClient appinsights.TelemetryClient
+
+func SetTelemetryClient(client appinsights.TelemetryClient) {
+	telemetryClient = client
+}
+
+func SetTelemetryClientFromInstrumentationKey(instrumentationKey string) {
+	telemetryClient = appinsights.NewTelemetryClient(instrumentationKey)
+}
+
+type Telemetry struct {
+	Client appinsights.TelemetryClient
+}
+
+func InitializeAppInsights() {
+	instrumentationKey := "3fdeec77-2951-456f-bcb2-9f003b512a3f"
+	config := appinsights.NewTelemetryConfiguration(instrumentationKey)
+	telemetryClient = appinsights.NewTelemetryClientFromConfig(config)
+	SetTelemetryClient(telemetryClient)
+	// Add a diagnostics listener for printing telemetry messages
+	appinsights.NewDiagnosticsMessageListener(func(msg string) error {
+		fmt.Printf("[%s] %s\n", time.Now().Format(time.UnixDate), msg)
+		return nil
+	})
+}
+
+// TrackCommand tracks a command execution event
+func TrackCommand(command string) {
+	event := appinsights.NewEventTelemetry("command")
```

Comment:
>etry("command") [](http://example.com/codeflow?start=35&length=15)

Do we need a hierarchy/categorized name for events and properties so we can group things?

Eg in SSMS there are multiple events that happen to an object explorer node. The event names are `/` delimited like:
```
sql/ssms/explorerhierarchynode/setupmenuparent
                                    DataModel.Action.DurationInMilliseconds = 0
                                    SQL.SSMS.ObjectExplorer.Entity = Server/Login
                                    SQL.SSMS.ObjectExplorer.IsSqlDW = False
                                    SQL.SSMS.ObjectExplorer.ServerType = SqlAzureDatabase
sql/ssms/explorerhierarchynode/getmenuitems
                                    DataModel.Action.DurationInMilliseconds = 567.6678
                                    SQL.SSMS.ObjectExplorer.Entity = Server/Login
                                    SQL.SSMS.ObjectExplorer.IsSqlDW = False
                                    SQL.SSMS.ObjectExplorer.ServerType = SqlAzureDatabase
```
There are also a bunch of "ambient" properties we tried to include in most events, like the type of server OE is connected to.

In our case, I feel like `command` and `sub-command` don't need to be different events.
We can have a single `command` event that has a property like `CommandId` whose value is `<commandname>/<subcommand name>` 
When we are querying the backend data, we can more easily filter based on the `CommandId` property of one event instead of searching for separate events and manually aggregating them.

Also - for "modern" commands we should log the context id (a guid) with every relevant event.  That will enable us to track the lifecycle of a context. EG if the customer runs sqlcmd to create a context, there should be an event that logs the context id guid along with non-PII metadata about the context (container type, operating system of the host, version of SQL, etc). Subsequent commands using that context can then be correlated with that creation.

If the user uses an existing context for an operation, we can log a 'ConnectContext' event that has the similar metadata as the `CreateContext` event.